src/wasm/wasm-ir-builder.cpp

```diff
@@ -458,18 +461,20 @@ struct IRBuilder::ChildPopper
     // We have checked all the constraints, so we are ready to pop children.
     for (int i = children.size() - 1; i >= 0; --i) {
       if (needUnreachableFallback &&
-          scope.exprStack.size() == *unreachableIndex + 1) {
-        // The expressions remaining on the stack may be executed, but they do
-        // not satisfy the requirements to be children of the current parent.
-        // Explicitly drop them so they will still be executed for their side
-        // effects and so the remaining children will be filled with
-        // unreachables.
-        assert(scope.exprStack.back()->type == Type::unreachable);
-        for (auto& expr : scope.exprStack) {
-          expr = Builder(builder.wasm).dropIfConcretelyTyped(expr);
-        }
+          scope.exprStack.size() == *unreachableIndex + 1 && i > 0) {
+        // The next item on the stack is the unreachable instruction we must
+        // not pop past. We cannot insert unreachables in front of it because
+        // it might be a branch we actually have to execute, so this next item
+        // must be child 0. But we are not ready to pop child 0 yet, so
```

Comment:
I am having a hard time following this. "The next item is the unreachable we must not pop past" fits `scope.exprStack.size() == *unreachableIndex + 1`. Then "this next item be child 0" seems to refer to the same "next item", which suggests `*unreachableIndex == 0`. Is that right? If so then I can't see why.

Reply:
No, `*unreachableIndex` is an index in the stack, not in the children we are popping. `i` is the index of the child we want to pop, so that's why we check `i > 0` here.

Reply:
That helps, thanks. But if `i` is the index of the child we want to pop, and we check `i > 0` before getting here, how can `so this next item must be child 0` be true?

Reply:
Because inside this condition we are _not_ popping a value from the stack, we are synthesizing an `unreachable` instead. We're deferring popping that value until `i == 0`.

Reply:
I see. Ok, I think I understand the surrounding code a bit better now. But I am still quite confused :smile:  Specifically, about this sentence:

> We cannot insert unreachables in front of it because it might be a branch we actually have to execute, so this next item must be child 0.

I can't follow the inference of "so". I can try to paraphrase the key part of it as "We can't insert unreachables, as it might be a branch, and therefore it must be child 0". Neither "we can't insert unreachables" nor "it might be a branch" suggest to me any reason why the child must be 0 and not any other index.

Reply:
We need to fill out all the children, ideally by popping from the stack. But we're in a situation where we can't fill all the children by popping from the stack either because the stack does not have enough values or because some type constraint would be violated. We definitely need to pop everything above the last unreachable-typed instruction in the stack, but we can't pop anything below that instruction. So after popping everything above the last unreachable instruction, we still have some number of children to fill and only one more instruction (the last unreachable one) that we are allowed to pop. The code previously popped that last unreachable instruction to become the latest possible child and filled the prior children with synthesized `Unreachable` instructions, but that was incorrect because it was effectively inserting `Unreachable` before what might be some kind of branch, causing the parsed program to trap instead of taking the branch. The code now fills children with synthesized `Unreachable` instructions first and waits until the end to pop that last unreachable instruction so that it becomes the very first child, which will be executed before any of the synthesized `Unreachable`s are reached.

Reply:
Thanks, I think I follow all of that, and it makes sense. But why is the child at index 0 special? I feel like I'm missing something very specific there that must follow from all that you said, but I can't guess what it is. To put it another way, why is child 1 never in the special situation that child 0 can be?

Reply:
If we put the unreachable instruction from the stack in the child 1 slot, then there would be nothing valid to do for the child 0 slot. Popping off the stack would be popping past the unreachable, which we determined we cannot do. Synthesizing an `Unreachable` for child 0 would insert that `Unreachable` before the unreachable instruction from the stack in the execution order and change the program semantics if the instruction from the stack was a branch.

Reply:
Why is that not a problem for the child 2 slot as well? If we put the unreachable instruction from the stack in the child 2 slot then there would be nothing to do for the child 1 or child 0 slots, I think, for the exact reasons you said?

Reply:
Exactly, it's a problem for any slot except for child 0. That's why the child 0 slot is special.

Reply:
I see now, thanks... sorry for being dense!

Reply:
Haha no worries! It's certainly a dense piece of code.